R/merge_sce_list.R

```diff
@@ -0,0 +1,199 @@
+#' @param shared_features A vector of features (genes) that all SCEs to be merged
+#'   have in common
+#' @param retain_coldata_cols A vector of columns to retain in the colData slot
+#' @param preserve_rowdata_cols A vector of rowData columns which should not be
+#'   renamed
+#' @param expected_coldata_names A vector of column names that are expected to be
+#'   present in the SCE. Any missing columns will be created with `NA` values.
+#'
+#' @return An updated SCE that is prepared for merging
+prepare_sce_for_merge <- function(sce,
+                                  sce_name,
+                                  batch_column,
+                                  barcode_column,
+                                  shared_features,
+                                  retain_coldata_cols,
+                                  preserve_rowdata_cols,
+                                  expected_coldata_names) {
+
+  # Current functionality does not retain any present altExps
+  sce <- removeAltExps(sce)
+
+  # Subset to shared features
+  sce <- sce[shared_features,]
+
+  ##### rowData #####
+  # Add `sce_name` ID to rowData column names except for those
+  #  present in `preserve_rowdata_cols`
+  original_colnames <- colnames(rowData(sce))
+
+  colnames(rowData(sce)) <- ifelse(
+    original_colnames %in% preserve_rowdata_cols,
+    original_colnames,
+    glue::glue("{original_colnames}-{sce_name}")
+  )
+
+  ##### colData #####
+  observed_coldata_names <- names(colData(sce))
+
+  # If the barcode column is not already present, create it
+  if (!(barcode_column %in% observed_coldata_names)) {
+    colData(sce)[,barcode_column] <- rownames(colData(sce))
+    observed_coldata_names <- c(observed_coldata_names, barcode_column)
+  }
+
+  # Ensure barcode_column is expected and retained
+  if (!(barcode_column %in% expected_coldata_names)) {
+    expected_coldata_names <- c(expected_coldata_names, barcode_column)
+  }
+  if (!(barcode_column %in% retain_coldata_cols)) {
+    retain_coldata_cols <- c(retain_coldata_cols, barcode_column)
+  }
+
+  # Ensure all columns are present in all SCEs by adding `NA` columns as needed
+  missing_columns <- setdiff(expected_coldata_names, observed_coldata_names)
+  for (missing_col in missing_columns) {
+    # Create the missing column only if it should be retained
+    if (missing_col %in% retain_coldata_cols) {
+      colData(sce)[[missing_col]] <- NA
+    }
+  }
+
+  # Retain only the columns present in `retain_coldata_cols`
+  colData(sce) <- colData(sce) |>
+    as.data.frame() |>
+    dplyr::select( dplyr::all_of(retain_coldata_cols) ) |>
+    S4Vectors::DataFrame()
+  # this approach doesn't work when retain is length 1:
+  #colData(sce) <- colData(sce)[, retain_coldata_cols]
+
+  # Add `sce_name` to colData row names so cell barcodes can be mapped to originating SCE
+  original_rownames <- rownames(colData(sce))
+  rownames(colData(sce)) <- glue::glue("{rownames(colData(sce))}-{sce_name}")
```

Comment:
I think we want to just update the whole SCE here?
I don't think you use `original_rownames` so I dropped it.

For organization, I might move this to be the first step of the `prepare` function, as it is one of the most important parts...

```suggestion
  # Add `sce_name` to colnames so cell ids can be mapped to originating SCE
  colnames(sce) <- glue::glue("{colnames(sce)}-{sce_name}")
```